=head1 NAME

joinx - a tool for performing set operations on .bed files

=head1 SYNOPSIS

joinx [--version|-v] SUBCOMMAND [SUBCOMMAND-ARGS] 

=head1 DESCRIPTION

Joinx is a tool for performing set operations (e.g., set intersection,
set difference, ...) on genomic data contained in .bed files. It also
provides some limited analysis functions (concordance reports).

=head1 INTERSECT SUBCOMMAND

joinx intersect
    Perform set intersection and difference operations.

=head2 SYNOPSIS

joinx intersect [OPTIONS] a.bed b.bed

=head2 OPTIONS

-h, --help
    Display usage information

-a, --file-a <path> (or first positional argument)
    Path to sorted input .bed file "A" (- for stdin)
    
-b, --file-b <path> (or second positional argument)
    Path to sorted input .bed file "B" (- for stdin, note: not both "A"
    and "B" can be stdin)

-o, --output-file <path>
    Output bed file for intersection results

--miss-a <path>
    Write entries in A that are not in B (set difference, A\B) to <path>.
    
--miss-b <path>
    Write entries in B that are not in A (set difference, B\A) to <path>.

-f, --first-only
    Notice only the first thing in A to hit records in B, not the full
    intersection

--output-both
    Concatenate the columns of A and B in the intersection output

--exact-pos
    For indels, define intersection to mean an exact position match. The
    default considers any overlap a match.

--exact-allele
    Require exact allele matching for intersection.

--iub-match
    When using exact-allele, treat any overlap between an allele and an
    IUB code as a match (e.g., G & K match).

--dbsnp-match
    Undocumented. Should not be used.
    
=head1 SNV-CONCORDANCE SUBCOMMAND

joinx snv-concordance
    Generate basic concordance reports given 2 .bed files.

=head2 SYNOPSIS

joinx snv-concordance [OPTIONS] a.bed b.bed

=head2 OPTIONS
-h, --help
    Display usage information

-a, --file-a <path> (or first positional argument)
    Path to sorted input .bed file "A" (- for stdin)
    
-b, --file-b <path> (or second positional argument)
    Path to sorted input .bed file "B" (- for stdin, note: not both "A"
    and "B" can be stdin)

-d, --depth
    Use read depth (bed column 5) instead of quality (column 4) in report

--hits <path>
    Output the resulting intersection of "A" and "B" to "path" (discarded by default)

--miss-a <path>
    Write entries in A that are not in B (set difference, A\B) to <path>.
    
--miss-b <path>
    Write entries in B that are not in A (set difference, B\A) to <path>.

=head1 SORT SUBCOMMAND

joinx sort
    Sort .bed files by chromosome/sequence, start, end position.

=head2 SYNOPSIS

joinx sort [OPTIONS] a.bed b.bed c.bed ... 

=head2 OPTIONS

-h, --help
    Display usage information

-i, --input-file <path>
    Input file to be sorted (can be specified multiple times, positional arguments work too)

-m, --merge-only
    Assume all input files are pre-sorted, merge only

-o, --output-file
    The path of the sorted output file to write

-M, --max-mem-lines (default=1000000)
    The maximum number of lines to hold in memory while sorting (before writing to tmp files)

-s, --stable
    Perform a "stable" sort (equivalent records maintain their original order)

=head1 AUTHOR

Joinx was written by Travis Abbott <tabbott@genome.wustl.edu>, and is maintained on
Github at: L<https://github.com/genome/joinx/>.

=head1 COPYRIGHT

Joinx is free software, distributed under the terms of the GNU GPL v3 or later: L<http://gnu.org/licenses/gpl.html>.

